linter: Ban shebang lines in files ending with '.py'.

derAnfaenger opened the pull request:
This creates ~45 linter conflicts. I think most of the files can be simply renamed, but there might be some instances where another piece of code contains their filename or uses them as a module.

Coming from #6872 :
> Let's leave that issue in the code-style guide

If we lint for it, shouldn't we be able to remove it?

timabbott commented:
I think this was somewhat ill-conceived, in that we can't lint for this everwhere, since some of these actually should have the shebang and end with `.py`, but I think all the cases in `zerver/` are bugs.  Maybe just do an `include_only` on the zerver/ directory?

rht commented:
The result of `grep -l '/usr/bin/env python'  | grep \\.py` 

```
manage.py
tools/run-dev.py
tools/test-help-documentation.py
docs/html_unescape.py
tools/send_github_payloads.py
tools/show-profile-results.py
tools/js-dep-visualizer.py
tools/lister.py
scripts/lib/hash_reqs.py
scripts/lib/zulip_tools.py
scripts/lib/queue_workers.py
scripts/setup/generate_secrets.py
zerver/views/email_log.py
tools/lib/provision.py
tools/lib/sanity_check.py
tools/linter_lib/custom_check.py
tools/setup/setup_venvs.py
zerver/management/commands/deliver_email.py
zerver/management/commands/email_mirror.py
zerver/management/commands/print_email_delivery_backlog.py
tools/documentation_crawler/documentation_crawler/spiders/check_help_documentation.py
tools/documentation_crawler/documentation_crawler/spiders/check_documentation.py
puppet/zulip/files/postgresql/pg_backup_and_purge.py
tools/documentation_crawler/documentation_crawler/spiders/common/spiders.py
```

In my opinion, with the exception of `manage.py`, the `.py` can be removed and then have the lint be enforced.

rht commented:
Also, `zerver/management/commands/*` are not supposed to be executable.

timabbott commented:
We can't do that for all of these; some of them need to be a `.py` file so that they can be imported by a test or other library code.  Certainly that's the case for `scripts/lib/queue_workers.py` and `lister.py`.  These ones definitely are bugs though:

Should not have a shebang:
* `zerver/views/email_log.py`
* `tools/documentation_crawler/documentation_crawler/spiders/common/spiders.py`
* `tools/lib/sanity_check.py`
* `tools/documentation_crawler/documentation_crawler/spiders/check_help_documentation.py`
* `tools/documentation_crawler/documentation_crawler/spiders/check_documentation.py`

Should be deleted:
* `html_unescape.py` (I just dealt)

Should remove `.py` extension:
* `tools/test-help-documentation.py`
* `tools/show-profile-results.py
* `tools/send_github_payloads.py`

We can start with those and rerun the report once resolved.

derAnfaenger commented:
@timabbott updated.

* updated the linter rule to only include `zerver`. `special_rules` is needed since directly adding the rule to `rules` results in a funny bug where the rule gets repeatedly added for every `.py` file.

* removed the shebangs and `.py` extensions in your list.

* added some other commits i found to improve consistency.

timabbott commented:
Merged everything but `special_rules`.  I think for that, I'd rather have a `shebang_rules` feature in the linter framework that's more specialized to this use case (e.g. it should only need to inspect the first line of the file!).

derAnfaenger commented:
@timabbott updated. I hope the new structure is about what you had in mind. It doesn't give the prettiest output, but is efficient and makes a good fit between firstline and lastline rules.

timabbott commented:
Merged, after fixing a few things:
* Using print rather than `print_err` (didn't have color, marking of which linter was involved).  We should probably add a lint rule for this or something :)
* Tweaked some of the error text.

Thanks @derAnfaenger!
